tests/unit/s2n_protocol_version_getter_test.c

```diff
@@ -0,0 +1,395 @@
+
+                uint8_t actual_protocol_version = s2n_connection_get_actual_protocol_version(server);
+                if (server_version == S2N_TLS12) {
+                    /* For backwards compatibility, TLS 1.2 servers always use the client hello
+                     * version to determine the client's maximum version, even if a supported
+                     * versions extension was received.
+                     */
+                    EXPECT_EQUAL(actual_protocol_version, MIN(server_version, client_hello_version));
+                } else {
+                    /* TLS 1.3 servers always use the version in the supported versions extension,
+                     * regardless of the client hello version.
+                     */
+                    EXPECT_EQUAL(actual_protocol_version, MIN(server_version, client_supported_version));
+                }
+            }
+        }
+    }
+
+    /* Test protocol version getters on the server when a supported versions extension isn't received */
+    for (uint8_t server_version = S2N_TLS12; server_version <= S2N_TLS13; server_version++) {
+        for (uint8_t client_hello_version = S2N_SSLv3; client_hello_version <= S2N_TLS12; client_hello_version++) {
+            DEFER_CLEANUP(struct s2n_config *client_config = s2n_config_new(), s2n_config_ptr_free);
+
+            /* A TLS 1.2 security policy is set to prevent the client from sending a supported
+             * versions extension.
+             */
+            EXPECT_SUCCESS(s2n_config_set_cipher_preferences(client_config, "test_all_tls12"));
+
+            DEFER_CLEANUP(struct s2n_config *server_config = s2n_config_new(), s2n_config_ptr_free);
+            EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(server_config, chain_and_key));
+
+            if (server_version == S2N_TLS12) {
+                EXPECT_SUCCESS(s2n_config_set_cipher_preferences(server_config, "test_all_tls12"));
+            } else {
+                if (!s2n_is_tls13_fully_supported()) {
+                    continue;
+                }
+                EXPECT_SUCCESS(s2n_config_set_cipher_preferences(server_config, "test_all"));
+            }
+
+            DEFER_CLEANUP(struct s2n_connection *client = s2n_connection_new(S2N_CLIENT),
+                    s2n_connection_ptr_free);
+            EXPECT_NOT_NULL(client);
+            EXPECT_SUCCESS(s2n_connection_set_config(client, client_config));
+
+            DEFER_CLEANUP(struct s2n_connection *server = s2n_connection_new(S2N_SERVER),
+                    s2n_connection_ptr_free);
+            EXPECT_NOT_NULL(server);
+            EXPECT_SUCCESS(s2n_connection_set_config(server, server_config));
+
+            struct s2n_stuffer *hello_stuffer = &client->handshake.io;
+            EXPECT_SUCCESS(s2n_client_hello_send(client));
+
+            /* Overwrite the client hello version according to the test case. */
```

Comment:
Maybe you could make this overwrite a function? Would make the tests easier to read.